src/slurm_plugin/task_executor.py

```diff
@@ -0,0 +1,51 @@
+import inspect
+import logging
+from concurrent.futures import ThreadPoolExecutor
+from threading import Semaphore
+
+logger = logging.getLogger(__name__)
+
+
+class TaskExecutor:
+    class MaximumBacklogExceededException(RuntimeError):
+        """Exception raised when a task can't be queued due to backlog."""
+
+        def __init__(self, task, maximum_backlog):
+            self.failed_task = task
+            self.maximum_backlog = maximum_backlog
+
+    """Class for managing execution of asynchronous tasks."""
+
+    def __init__(self, worker_pool_size, max_backlog):
+        self._max_backlog = max_backlog
+        self._executor_limit = Semaphore(max_backlog)
+        self._executor_pool = ThreadPoolExecutor(max_workers=worker_pool_size)
```

Comment:
Yes, it will hang until all busy worker threads complete - for Python versions >= 3.9, we pass a flag to cancel any pending tasks, but versions < 3.9 do not support that option and so the pool will still complete all pending tasks in the queue before letting the process exit.

Reply:
having the clusetermgtd hang for 3 minutes in the worst case scenario is not a good idea. I'm wondering if we should and could just forcefully kill all threads, or eventually give them a very very small time to complete (e.g. 2secs).

Reply:
I've refactored a bit - so now the tasks will wait on what is ultimately an event. When a shutdown is requested, the event will be set and any "sleeping" tasks will raise an exception. If someone cares about the success or failure of the task, the returned Future can be checked for task status - either cancelled or that an exception was raised.